hyperspy/drawing/marker.py

```diff
@@ -0,0 +1,296 @@
+    add_data
+        Add data to the structured array. Same field as set_data
+
+    Example
+    -------
+
+    >>> s = signals.Spectrum(random.random([10, 100]))
+    >>> m = utils.plot.marker('axvline')
+    >>> m.set_marker_properties(color='green')
+    >>> m.set_data(x1=range(10))
+    >>> s.plot()
+    >>> s._plot.signal_plot.add_marker(m)
+    >>> m.plot()
+
+    >>> im = signals.Image(random.random([10, 50, 50]))
+    >>> m = utils.plot.marker('text')
+    >>> m.set_marker_properties(fontsize = 30, color='red')
+    >>> m.set_data(x1=range(10), y1=range(10)[::-1], text='hello')
+    >>> im.plot()
+    >>> im._plot.signal_plot.add_marker(m)
+    >>> m.plot()
+
+    >>> im = signals.Image(np.zeros((100, 100)))
+    >>> m = utils.plot.marker('rect')
+    >>> m.set_marker_properties(linewidth=4, color='red', linestyle='dotted')
+    >>> m.set_data(x1=20, x2=70, y1=20, y2=70)
+    >>> im.plot()
+    >>> im._plot.signal_plot.add_marker(m)
+    >>> m.plot()
+
+    """
+
+    def __init__(self, marker_type='line'):
+        # Data attributes
+        self.data = None
+        self.axes_manager = None
+        self.ax = None
+        self.auto_update = True
+
+        # Properties
+        self.marker = None
+        self.orientation = None
+        self._marker_properties = {}
+        self.type = marker_type
+
+    @property
+    def type(self):
+        return self._type
+
+    @type.setter
+    def type(self, value):
+        lp = {}
+        if value == 'text':
+            lp['color'] = 'black'
+        elif value == 'line':
+            lp['linewidth'] = 1
+            lp['color'] = 'black'
+        elif value == 'axvline':
+            lp['linewidth'] = 1
+            lp['color'] = 'black'
+        elif value == 'axhline':
+            lp['linewidth'] = 1
+            lp['color'] = 'black'
+        elif value == 'pointer':
+            lp['color'] = 'black'
+            lp['linewidth'] = None
+        elif value == 'rect':
+            lp['color'] = 'black'
+            lp['fill'] = None
+            lp['linewidth'] = 1
+        else:
+            raise ValueError(
+                "`type` must be one of {\'line\', "
+                "\'axvline\', \'axhline\', \'text\', \'pointer\', \'rect\'}"
+                "but %s was given" % value)
+        self._type = value
+        self.marker_properties = lp
+
+    @property
+    def marker_properties(self):
+        return self._marker_properties
+
+    @marker_properties.setter
+    def marker_properties(self, kwargs):
+        if 'type' in kwargs:
+            self.type = kwargs['type']
+            del kwargs['type']
+
+        for key, item in kwargs.iteritems():
+            if item is None and key in self._marker_properties:
+                del self._marker_properties[key]
+            else:
+                self._marker_properties[key] = item
+        if self.marker is not None:
+            plt.setp(self.marker, **self.marker_properties)
+            if self.ax is None:
```

Comment:
I do not thing that is possible - i.e. if the `self.marker` is not `None`, then you can actually get the `self.ax` with `self.marker.axis`. The `self.marker` cannot be created to begin with, if the `self.ax` is not set - hence I think the checking should be done before creating it (my guess would be at the very start of `plot` ).

The warning could also be improved - at the moment it would require the user to figure out the structure of both hyperspy plotting (difficult!), matplotlib, and the marker itself to understand what the `self.ax` refers to. I think adding a user-friendly sentence (e.g. "please add marker to a plot with `[..some things..].add_marker()`") should help.

Let me know if that does not make sense